sdk/eventgrid/eventgrid/src/cloudEventDistrubtedTracingEnricherPolicy.ts

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import {
+  HttpOperationResponse,
+  RequestPolicy,
+  RequestPolicyOptions,
+  RequestPolicyFactory,
+  WebResourceLike
+} from "@azure/core-http";
+
+export const CloudEventBatchContentType = "application/cloudevents-batch+json; charset=utf-8";
+export const TraceParentHeaderName = "traceparent";
+export const TraceStateHeaderName = "tracestate";
+export const ContentTypeHeaderName = "Content-Type";
+
+export function cloudEventDistributedTracingEnricherPolicy(): RequestPolicyFactory {
+  return {
+    create: (nextPolicy: RequestPolicy, _options: RequestPolicyOptions) => {
+      return new CloudEventDistributedTracingEnricherPolicy(nextPolicy);
+    }
+  };
+}
+
+/**
+ * CloudEventDistributedTracingEnricherPolicy is a policy which adds distributed tracing informatiom
+ * to a batch of cloud events. It does so by copying the `traceparent` and `tracestate` properties
+ * from the HTTP request into the individual events as extension properties.
+ *
+ * This will only happen in the case where an event does not have a `traceparent` defined already. This
+ * allows events to explicitly set a traceparent and tracestate which would be respected during "multi-hop
+ * transmition".
+ *
+ * See https://github.com/cloudevents/spec/blob/master/extensions/distributed-tracing.md
+ * for more information on distributed tracing and cloud events.
+ */
+export class CloudEventDistributedTracingEnricherPolicy implements RequestPolicy {
+  constructor(readonly _nextPolicy: RequestPolicy) {}
+
+  sendRequest(httpRequest: WebResourceLike): Promise<HttpOperationResponse> {
+    if (!httpRequest) {
+      throw new Error("webResource cannot be null or undefined");
+    }
+
+    const traceparent = httpRequest.headers.get(TraceParentHeaderName);
+    const tracestate = httpRequest.headers.get(TraceStateHeaderName);
+
+    if (
+      httpRequest.headers.get(ContentTypeHeaderName) === CloudEventBatchContentType &&
+      typeof httpRequest.body === "string" &&
+      traceparent
+    ) {
+      // per the cloud event batched content type we know the body is an array encoded in JSON.
+      const body = JSON.parse(httpRequest.body) as any[];
```

Comment:
We really need to figure out how to avoid this double serialization cost soon... but I don't think there is a good way around it right now without us having to teach `AutoRest.TypeScript` about including the header on the mapped object form unless @joheredi knows of any mapper tricks.

Reply:
After looking at Jose's recent PR for the batch stuff in tables, I wonder if something like setting `disableJsonStringifyOnBody` on the request options when calling the client method would work, and then just expect that this policy (which would be at the end of the chain) would always do the JSON.stringify (after enriching the events if needed)?  Does that seem like a reasonable strategy?

Reply:
`disableJsonStringifyOnBody` only works if you call `sendRequest` directly - it won't work from generated code that calls `sendOperationRequest`.

Reply:
Thanks, Jeff.  I'll stick with this approach for now and when we have a better story go clean things up.